websockets/websockets_test.go

```diff
@@ -879,3 +880,241 @@ func TestCookiesDefaultJar(t *testing.T) {
+				return
+			}
+		}))
+
+		err := ts.ev.Start(func() error {
+			_, runErr := ts.rt.RunString(sr(`
+		// if client supports compression, it has to send the header
+		// 'Sec-Websocket-Extensions:permessage-deflate; server_no_context_takeover; client_no_context_takeover' to server.
+		// if compression is negotiated successfully, server will reply with header
+		// 'Sec-Websocket-Extensions:permessage-deflate; server_no_context_takeover; client_no_context_takeover'
+
+		var params = {
+			"compression": "deflate"
+		}
+		var ws = new WebSocket("WSBIN_URL/ws-compression", null, params)
+		ws.onmessage = (data) => {
+			if (data != "` + text + `"){
+				throw new Error("wrong message received from server: ", data)
+			}
+
+			ws.close()
+		}
+		`))
+			return runErr
+		})
+
+		require.NoError(t, err)
+
+		samples := metrics.GetBufferedSamples(ts.samples)
+		assertSessionMetricsEmitted(t, samples, "", sr("WSBIN_URL/ws-compression"), http.StatusSwitchingProtocols, "")
+	})
+
+	t.Run("params", func(t *testing.T) {
+		t.Parallel()
+		testCases := []struct {
+			compression   string
+			expectedError string
+		}{
+			{
+				compression:   `""`,
+				expectedError: `unsupported compression algorithm '', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `null`,
+				expectedError: `unsupported compression algorithm 'null', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `undefined`,
+				expectedError: `unsupported compression algorithm 'undefined', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `" "`,
+				expectedError: `unsupported compression algorithm '', supported algorithm is 'deflate'`,
+			},
+			{compression: `"deflate"`},
+			{compression: `"deflate "`},
+			{
+				compression:   `"gzip"`,
+				expectedError: `unsupported compression algorithm 'gzip', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `"deflate, gzip"`,
+				expectedError: `unsupported compression algorithm 'deflate, gzip', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `"deflate, deflate"`,
+				expectedError: `unsupported compression algorithm 'deflate, deflate', supported algorithm is 'deflate'`,
+			},
+			{
+				compression:   `"deflate, "`,
+				expectedError: `unsupported compression algorithm 'deflate,', supported algorithm is 'deflate'`,
+			},
+		}
+
+		for _, testCase := range testCases {
+			testCase := testCase
+			t.Run(testCase.compression, func(t *testing.T) {
+				t.Parallel()
+				ts := newTestState(t)
+				sr := ts.tb.Replacer.Replace
+				ts.tb.Mux.HandleFunc("/ws-compression-param", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					upgrader := websocket.Upgrader{
+						EnableCompression: true,
+						ReadBufferSize:    1024,
+						WriteBufferSize:   1024,
+					}
+
+					conn, e := upgrader.Upgrade(w, req, w.Header())
+					if e != nil {
+						t.Fatalf("/ws-compression-param cannot upgrade request: %v", e)
+						return
+					}
+
```

Comment:
nit: I feel like you can probably extract this handler and in this place have a function call that does the per case action you need. (Like writing a message back)